design_docs/generalize_calendar.md

```diff
@@ -0,0 +1,303 @@
+timescales) with this system.
+
+For purposes of performing mathematical operations and plotting dates, these
+values will be converted to `datetime.datetime` objects (via the proposed
+`days_to_datetime` utility function) and back (via the proposed 
+`datetime_to_days`).
+
+The conversion operations within `datetime_to_days` and `days_to_datetime` will be
+performed with the calendar-aware functions `netCDF4.date2num` and 
+`netCDF4.num2date`, respectively.  Both functions will support lists/arrays of dates
+(for efficiency and simplicity of calling code) in addition to single values.
+
+Curve ploting can be supported with `matplotlib.pyplot.plot_date`, which takes a date
+of exactly the format used here (days since 0001-01-01).  The compatibility with `plot_date`
+was part of the reason for choosing this format for the date.
+
+<h2> Design solution: The 'Time' coordinate of xarray data sets must support at least years 
+0001 and 9999, and preferably any conceivable value<br>
+Date last modified: 2017/02/09 <br>
+Contributors: Xylar Asay-Davis
+</h2>
+
+Same as above. In theory, the use of days since 0001-01-01 would allow any year
+to be supported, not just the range from 0001 to 9999.  However, the conversions
+to `datetime.datetime` objects for mathematical manipulation will constrain
+the dates to be between `datetime.min` (0001-01-01) and `datetime.max` (9999-12-31).
+
+
+<h1> Design and Implementation </h1>
+
+<h2> Implementation: The 'Time' coordinate of xarray data sets must be consistent
+with the MPAS calendar <br>
+Date last modified: 2017/02/16 <br>
+Contributors: Xylar Asay-Davis
+</h2>
+
+The proposed implementation is on the branch 
+[xylar/generalize_calendar](https://github.com/xylar/MPAS-Analysis/tree/generalize_calendar)
+
+A helper funciton, `mpas_xarray._parse_dataset_time`, computes times as days since 
+0001-01-01, and serves as a replacement for `mpas_xarray._get_datetimes`. 
+
+**Note: the current implementation breaks the convention that `mpas_xarray` remains
+separate from the rest of MPAS-Analyis by using 3 functions from `timekeeping.utility`
+in `mpas_xarray`:**
+```python
+from ..timekeeping.utility import string_to_days_since_date, \
+    days_to_datetime, datetime_to_days
+```
+**This violates the first requirement in the 
+[Design Document: Moving variable mapping out of mpas_xarray](https://github.com/xylar/MPAS-Analysis/blob/design_doc_variable_mapping_reorg/design_docs/variable_mapping_reorg.md).
+I am open to alternative solutions for keeping `mpas_xarray` separate from the rest
+of analysis but these 3 functions do not conceptually belong in `mpas_xarray`.  The
+problem is exacerbated by the fact that there are analysis-specific functions in
+`timekeeping`, meaning that this cannot easily be made a submodule of `mpas_xarray`
+(nor would this make very much logical sense).  Having 2 `timekeeping` modules, one
+for `mpas_xarray` and one for MPAS-Analysis, seems unnecessarily confunsing.**
```

Comment:
@xylar, why do we need to have the utilities associated with `mpas_xarray` in the `timekeeping.utility` file?  As long as there are not a large number of these it is ok to include them in `mpas_xarray`- this is the simplest solution and resolution to the problem.

The solution of having two timekeeping modules will be confusing long term and I vote against it.

We are at the point, e.g., timekeeping in xarray isn't robust enough to handle MPAS needs, that it may be both impractical and unreasonable to keep MPAS-Analysis and mpas_xarray separate.  The whole point of mpas_xarray is to provide an xarray-compatable dataset that can leverages dask for performance.  We can't fully do this without handling the troubling time issue, which admittedly is even a fix for the MPAS time datastructures.  Hence, we are at the point where keeping mpas_xarray and MPAS-Analysis separate makes less sense and potentially may be confusing in the long term too.  This being said, once time is fixed in xarray, we should consider doing a complete refractor (at least in terms of xarray interfacing).